x/wait_stream/writer.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wait_stream
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/Jigsaw-Code/outline-sdk/x/sockopt"
+)
+
+type waitStreamWriter struct {
+	conn       io.Writer
+	tcpOptions sockopt.TCPOptions
+}
+
+var _ io.Writer = (*waitStreamWriter)(nil)
+
+func NewWriter(conn io.Writer, tcpOptions sockopt.TCPOptions) io.Writer {
+	return &waitStreamWriter{
+		conn:       conn,
+		tcpOptions: tcpOptions,
+	}
+}
+
+func (w *waitStreamWriter) Write(data []byte) (written int, err error) {
+	written, err = w.conn.Write(data)
+
+	// This may not be implemented, so it's best effort really.
+	waitUntilBytesAreSentErr := w.tcpOptions.WaitUntilBytesAreSent()
```

Comment:
I think this should be before the Write is done.

Also, this may significantly slow down communication. What's the performance impact? We will need to understand that for practical use.

We may want to restrict the wait for only when it's needed (on splits).

Reply:
Used a `fetch-speed` tool:
```
go run *.go -timeout 100500s -transport "waitstream"  http://speedtest.belwue.net/100M
Downloaded 100.00 MiB in 9.01s
Downloaded Speed: 11.10 MiB/s
```

Baseline:
```
go run *.go -timeout 100500s -transport ""  http://speedtest.belwue.net/100M                                                                                          
Downloaded 100.00 MiB in 9.75s
Downloaded Speed: 10.26 MiB/s
```